DDD-14 Proposal for new release process (#18)
debezium/debezium-design-documents: DDD-14.md

# DDD-14 Proposal for reformed release process

Current release process is more or less automation of original procedure defined in [RELEASING.md](https://github.com/debezium/debezium/blob/main/RELEASING.md).
The process uses staging approach.
Artifacts are built and uploaded into staging repository and then a smoke test is executed using those artifacts.
Over the time we added additional repositories into project and this approach gives us the notion of atomic releases - either artifacts from all repositories are released or none.
To achieve this quality the automated release process uses these traits

> * The release is executed twice, once as a dry run to verify that artifacts and be built and deployed and then the release itself

Reviewer: ```suggestion
The release is executed twice, once as a dry run to verify that artifacts are built and deployed and second time as the release itself
```

As part of publication the portal verifies a set of [requirements](https://central.sonatype.org/publish/requirements/) to which the bundle must conform.
One of them is the check of dependencies.
The issue is that if two repositories are deployed separately and depend on each other the depending one fails validation till referred one is published.
This breaks atomicity of releases.

As an alternative it is possible to build multiple repositories and upload them as a single bundle.
The problem is that the size of the bundle is limited to 1 GB.

The advantage of the new API is that the bundles can be published automatically by an API call.

## Problems with current release process
* Atomic releases are not possible
    * There is dependency between repositories which brekas validation
    * Debezium Server bundle can be very large so we can get near the size limit
* Process runs twice which adds to release execution time
* Human intervention is necessary

## Goal
* Make process compatible with new Maven portal
* Keep the dry run but use it for process testing only, not as a part of release

> * Remove the need of human intervention

Reviewer 2: I would assume this point comes back to something you mentioned in chat, and this is the idea for the release process to be resumeable/self healable, or at the very least having a way to clean-up a prior failed release before restarting it.

# Proposed release process

## Assumptions
* Jira is curated

> * Two PRs with documentation updates are available in core repository and documentation repository

Reviewer: what does documentation repository means, https://github.com/debezium/debezium.github.io/ ?

Reviewer 2: It is. I believe he's referring to the fact we need to modify the antora.yml file and changelog details in the main repository and then we need to construct a similar set of files in the website repository to include the release series, release data file, and the update to the release notes.

    * changelog, release notes, list of contributors, ...

## Release pipeline jobs

### Validation
* Documentation PRs exist and contains change for the given version
* Documentation PR is based on the latest `main` in core
* Jira
    * Release exists
    * All issues are `Resolved`/`Closed`

>     * All issues have `Component` set

Reviewer 2: These steps will eventually need to be based on GitHub and not Jira, so we should design the new process with that in mind.
